index.js

```diff
@@ -361,16 +349,16 @@ Router.prototype.off = function off (method, path) {
 }
 
 Router.prototype.lookup = function lookup (req, res, ctx) {
-  var handle = this.find(req.method, sanitizeUrl(req.url), this.versioning.deriveVersion(req, ctx))
+  var handle = this.find(req.method, sanitizeUrl(req.url), this.constrainer.deriveConstraints(req, ctx))
   if (handle === null) return this._defaultRoute(req, res, ctx)
   return ctx === undefined
     ? handle.handler(req, res, handle.params, handle.store)
     : handle.handler.call(ctx, req, res, handle.params, handle.store)
 }
 
-Router.prototype.find = function find (method, path, version) {
+Router.prototype.find = function find (method, path, derivedConstraints) {
```

Comment:
Why `derivedConstraints` must be defined?
Can't you assign it a default value?
```suggestion
Router.prototype.find = function find (method, path, derivedConstraints = {}) {
```
If we merge this change, the test should be updated as well.

Reply:
The `derivedConstraints` object is generated by `constrainer.deriveConstraints()` in `Router.lookup()`:
https://github.com/airhorns/find-my-way/blob/constraints/index.js#L352

And the `deriveConstraints` function is built [here](https://github.com/airhorns/find-my-way/blob/constraints/lib/constrainer.js#L74), and can return the value `undefined` if no constraint is in use.

Reply:
Sorry, not sure I follow, by looking at some test that you have updated, you can pass an empty object (and thus, use a default if the `derivedConstraints ` are not defined):
https://github.com/airhorns/find-my-way/blob/5344dd5fa2520016cbfb39dea1433f53a7a15cf5/test/issue-62.test.js#L15-L16

Reply:
So this was deliberate for a couple of reasons. First, I wanted to make the API really clear in that you should almost definitely be using `lookup` and not `find` because what you pass in for this `derivedConstraints` object really matters. We can default it to something, but I don't think we really want to make this API ergonomic, because if you don't pass it, you are just turning off the constraint system without really knowing. If the contract of the router is to match constrained routes to requests, that feels kind of dangerous that it would just turn off that constraining part. Especially if we're going to make this a breaking change, I feel like it's a good opportunity to change the API to reflect the actual needs of the function. If that's weird I am happy to change it but it expressed intent to me which I think is important.

Second reason is just to avoid the extra statement and/or extra allocation, prioritizing performance over ergonomics.

I also kind of think `find` I think should be private to be honest, but I was trying to preserve some measure of backwards compatibility.

Reply:
I would like to keep the `find` invocation free of additional params when constraints are not needed. There are use cases where find is more desirable compared to lookup.

Reply:
> I would like to keep the find invocation free of additional params when constraints are not needed. There are use cases where find is more desirable compared to lookup.

That means that the caller of `find` needs to be aware of if the router has constraints or not, which I think is kind of an anti-pattern. If someone is using `find-my-way` as part of a small system where they definitely know no routes are constrained, then what you're suggesting would make that a teensy bit easier, but is that really worth optimizing for? Anyone using `find-my-way` as part of a bigger ecosystem won't know what routes are registered ahead of time, in which case the API asking for constraint values is actually a positive as it forces them to implement the interface `find-my-way` needs to do it's job. 

Defaulting it also adds an extra object allocation which genuinely affects performance significantly, so we could do that if you're ok accepting the performance regression.

Reply:
> I'm lost. So passing findMyWay.find('GET', '/a/b') should work and disable all possible constraints?

It will work, correct, because it gets interpreted as `findMyWay.find('GET', '/a/b', undefined)`, which is what `.lookup` would pass when there are no constrained routes in use. The `undefined` there is what the constrainer produces when there aren't any constrained routes to avoid allocating an object or fishing out the version and host from the request. We needed to heavily optimize how the constrainer works to prevent virtually any extra work being done to support constraints when there are no routes to meet your folks' stringent performance requirements. Totally fine doing that, but we kind of had to which is why it is a bit of a strange API. 

> Constraints are niche use case and I think requiring this additional parameter for this is probably not needed. If undefined is a valid value, let's add a test to cover it and that'd be ok.

If you build a thing that uses find-my-way and doesn't pass a value for constraints, it is a bug in waiting though, isn't that ... bad? I think find-my-way really lends itself to being used as part of other resuable systems (like fastify, restify, etc), where the author of the reusable system can't necessarily guarantee what a different author might add as routes. @mcollina writes fastify, @airhorns writes routes, so if @airhorns decides to use constraints and @mcollina's fastify implementation doesn't pass the right stuff to find-my-way, the contract is broken. Why would we want to open up the API contract to that? 

I'll add the undefined test case, I think that's a great idea, but I think it'd be better to represent the 3rd parameter as a real thing that really should be passed if you're using `.find` in the docs and in the tests for ^ that reason.

Also, this should probably be a larger conversation and doesn't need to be solved now, but this problem would also go away if we privatized `.find` or changed it's API to be more useful. I think if you want a public`.find`, it should maybe take a request, not a set of constraints, and derive the constraints itself, as that would make this whole discussion moot. I am not really sure what the use cases for finding routes outside the context of a request is, especially if constraints are a thing, because you'd have to derive them yourself.

Reply:
I think we should just document:

1. that the 3rd argument is optional, i.e. `findMyWay.find('GET', '/a/b')` works.
2. using `findMyWay.find('GET', '/a/b', undefined)` would lead to _unspecified_ behavior if there are any constraints in the system.

Reply:
Ok, I'll do that!